File: virtual-world-simulator/src/Entity/Animals/Animal.cpp
#include "Animal.h"

int Animal::numberOfAnimals = 0;

Animal::Animal(BMPFile* texture, World* world, Point position, char symbol, int strength, int priority) :
	Entity{ texture, world, position, symbol, strength, priority }
{
    numberOfAnimals++;
}

Animal::~Animal()
{
    numberOfAnimals--;
}

void Animal::move(std::fstream& logFile, const Point& newPosition)
{
    Entity*** map = world->getMap();
    Entity* entity = map[newPosition.x][newPosition.y];
    if (entity != nullptr)
    {

        bool result = entity->collision(logFile, *this);

        if (this->isAlive == false)
        {
            map[position.x][position.y] = nullptr;
            return;
        }

        if (result)
        {
            if (entity->checkIfAlive() == false && entity->getPriority() > this->priority)
                delete entity;
        }
        else
        {
            return;
        }
    }

    map[position.x][position.y] = nullptr;
    this->position = newPosition;
    map[newPosition.x][newPosition.y] = this;
}

void Animal::update(std::fstream& logFile)
{
    int currentX = this->position.x;
    int currentY = this->position.y;
    int newX = currentX;
    int newY = currentY;
    
    while (newX == currentX && newY == currentY)
    {
        newX = currentX + Utility::random(-1, 1);
        newY = currentY + Utility::random(-1, 1);

        if (newX < 0)
            newX = 0;
        else if (newX > world->getHeight() - 1)
            newX -= 1;

        if (newY < 0)
            newY = 0;
        else if (newY > world->getWidth() - 1)
            newY -= 1;
    }

    this->move(logFile, { newX, newY });
	Entity::update(logFile);
}

bool Animal::collision(std::fstream& logFile, Entity& entity)
{
    //New animals are immortal in the first round
    if(this->lifespan == 0)
		return false;

    int entityStrength = entity.getStrength();

    if (entityStrength > this->strength ||
        (entityStrength == this->strength && entity.getLifespan() > this->lifespan))
    {
        logFile << entity.getPosition() << ' ' << this->symbol << " was killed by " << entity.getSymbol() << '\n';
        this->isAlive = false;
        return true;
    }

    logFile << entity.getPosition() << ' ' << entity.getSymbol() << " was killed by " << this->symbol << '\n';
    entity.kill();
    return false;
}

int Animal::getNumberOfAnimals()
{
    return numberOfAnimals;
}

File: Python/virtual-world-simulator/src/Entity/Animals/Animal.py
from src.Entity.Entity import Entity
import random


class Animal(Entity):

    def __init__(self, texture, world, position, symbol, strength, priority):
        super().__init__(texture, world, position, symbol, strength, priority)

    def move(self, newPosition):
        entity = self._world.getMapElement(newPosition[0], newPosition[1])
        if entity is not None:
            result = entity.collision(self)
            if self._isAlive is False or not result:
                return

        self._world.setMapElement(self._position[0], self._position[1], None)
        self._position = (newPosition[0], newPosition[1])
        self._world.setMapElement(self._position[0], self._position[1], self)

    def update(self):
        currentX = self._position[0]
        currentY = self._position[1]
        newX = currentX
        newY = currentY

        while newX == currentX and newY == currentY:
            newX = currentX + random.randint(-1, 1)
            newY = currentY + random.randint(-1, 1)

            if newX < 0:
                newX = 0
            elif newX > self._world.height - 1:
                newX -= 1

            if newY < 0:
                newY = 0
            elif newY > self._world.width - 1:
                newY -= 1

        self.move((newX, newY))
        self._lifespan += 1

    def collision(self, entity):
        if self._lifespan == 0:
            return False

        entityStrength = entity.strength

        if entityStrength > self._strength or (entityStrength == self._strength and entity.lifespan > self._lifespan):
            self._world.addLog(self._position, self._symbol + " was killed by " + entity.symbol)
            self.kill()
            return True

        self._world.addLog(self._position, entity.symbol + " was killed by " + self._symbol)
        entity.kill()
        return False

    def breed(self):
        currentX = self._position[0]
        currentY = self._position[1]
        freeFields = []

        for i in range(currentX - 1, currentX + 2):
            if i < 0 or i > self._world.height - 1:
                continue

            for j in range(currentY - 1, currentY + 2):
                if j < 0 or j > self._world.width - 1:
                    continue

                if self._world.getMapElement(i, j) is None:
                    freeFields.append((i, j))

        numberOfPossibleFields = len(freeFields)
        if numberOfPossibleFields > 0:
            newPosition = freeFields[random.randint(0, numberOfPossibleFields - 1)]
            self._world.addLog(newPosition, self._symbol + " born")
            return newPosition

